test-case/check-xrun-injection.sh

```diff
@@ -0,0 +1,117 @@
+#!/bin/bash
+
+##
+## Case Name: check-xrun-injection
+## Preconditions:
+##    N/A
+## Description:
+##    check xrun injection during playback/capture
+##    default duration is 10s
+##    default interval time of xrun injection is 0.5s
+## Case step:
+##    1. Parse TPLG file to get pipeline
+##    2. Specify the audio parameters
+##    3. Run aplay or arecord on each pipeline with parameters
+##    4. do xrun injection during playback or capture
+## Expect result:
+##    The return value of aplay/arecord should be 0
+##
+
+source $(dirname ${BASH_SOURCE[0]})/../case-lib/lib.sh
+
+OPT_OPT_lst['t']='tplg'     OPT_DESC_lst['t']='tplg file, default value is env TPLG: $TPLG'
+OPT_PARM_lst['t']=1         OPT_VALUE_lst['t']="$TPLG"
+
+OPT_OPT_lst['d']='duration' OPT_DESC_lst['d']='aplay/arecord duration in second'
+OPT_PARM_lst['d']=1         OPT_VALUE_lst['d']=10
+
+OPT_OPT_lst['i']='interval'     OPT_DESC_lst['i']='interval time of xrun injection'
+OPT_PARM_lst['i']=1         OPT_VALUE_lst['i']=0.5
+
+OPT_OPT_lst['p']='playback-file'   OPT_DESC_lst['p']='input file for playback'
+OPT_PARM_lst['p']=1         OPT_VALUE_lst['p']='/dev/zero'
+
+OPT_OPT_lst['c']='capture-file'   OPT_DESC_lst['c']='output file for catpure'
+OPT_PARM_lst['c']=1         OPT_VALUE_lst['c']='/dev/null'
+
+OPT_OPT_lst['s']='sof-logger'   OPT_DESC_lst['s']="Open sof-logger trace the data will store at $LOG_ROOT"
+OPT_PARM_lst['s']=0             OPT_VALUE_lst['s']=1
+
+func_opt_parse_option $*
+
+tplg=${OPT_VALUE_lst['t']}
+duration=${OPT_VALUE_lst['d']}
+interval=${OPT_VALUE_lst['i']}
+
+[[ ${OPT_VALUE_lst['s']} -eq 1 ]] && func_lib_start_log_collect
+
+func_lib_setup_kernel_last_line
+func_lib_check_sudo
+func_pipeline_export $tplg "type:playback,capture,both"
+
+declare -A CMD FILE
+CMD['playback,both']='aplay'
+FILE['playback,both']="${OPT_VALUE_lst['p']}"
+CMD['capture,both']='arecord'
+FILE['capture,both']="${OPT_VALUE_lst['c']}"
+
+xrun_injection()
+{
+    count=1
+    while(true)
+    do
+        ps -ef |grep "$pid" |grep -v grep
+        if [ $? -eq 0 ]; then
+            dlogc "XRUN injection: $count"
+            sudo bash -c "'echo 1 > /proc/asound/card0/$pcm/sub0/xrun_injection'"
+            sleep $interval
+            let count++
+        else
+            break # aplay/arecord is finished, stop xrun injection
+        fi
+    done
+}
+
+check_audio_file()
+{
+    if [[ ! -f ${FILE[$1]} ]]; then
+        dloge "${FILE[$1]} does not exist, skip the test."
+	return 1
```

Comment:
here have the function return

Reply:
Let's make it in an easy way. Do not accept user specified audio files, but use /dev/zero and /dev/null as default.